File: receiver_android/app/src/main/java/com/qrstream/receiver/protocol/ReceiverCore.kt
package com.qrstream.receiver.protocol

import org.json.JSONObject
import java.util.TreeMap
import java.util.zip.CRC32

class ReceiverCore {

    private var metadata: JSONObject? = null

    private val chunks = TreeMap<Int, ByteArray>()

    private var totalChunks = 0

    private var transferFinished = false

    // --------------------------------------------

    fun receive(packet: Packet) {

        verifyCRC(packet)

        when (packet.packetType) {

            PacketType.START -> receiveStart(packet)

            PacketType.DATA -> receiveData(packet)

            PacketType.END -> receiveEnd()

        }

    }

    // --------------------------------------------

    private fun verifyCRC(packet: Packet) {

        val crc = CRC32()

        crc.update(packet.payload)

        if (crc.value != packet.crc32) {

            throw RuntimeException("CRC Verification Failed")

        }

    }

    // --------------------------------------------

    private fun receiveStart(packet: Packet) {

        if (metadata != null) {
            return
        }

        metadata = JSONObject(String(packet.payload))

        totalChunks = metadata!!.getInt("total_chunks")

        chunks.clear()

        transferFinished = false

    }

    // --------------------------------------------

    private fun receiveData(packet: Packet) {

        // Ignore duplicate QR scans
        if (chunks.containsKey(packet.chunkNumber)) {
            return
        }

        chunks[packet.chunkNumber] = packet.payload

    }

    // --------------------------------------------

    private fun receiveEnd() {

        transferFinished = true

    }

    // --------------------------------------------

    fun completed(): Boolean {

        return transferFinished &&
                chunks.size == totalChunks

    }

    // --------------------------------------------

    fun getReceivedChunkCount(): Int {

        return chunks.size

    }

    // --------------------------------------------

    fun getTotalChunks(): Int {

        return totalChunks

    }

    // --------------------------------------------

    fun getMissingChunks(): List<Int> {

        val missing = mutableListOf<Int>()

        for (i in 0 until totalChunks) {

            if (!chunks.containsKey(i)) {

                missing.add(i)

            }

        }

        return missing

    }

    // --------------------------------------------

    fun getMetadata(): JSONObject? {

        return metadata

    }

    // --------------------------------------------

    fun getChunks(): Map<Int, ByteArray> {

        return chunks

    }

}

File: sender/receiver_core.py
"""
QRStream Receiver Core

Simulates the Android receiver.

Responsibilities:
- Read packets
- Verify CRC
- Store chunks
- Reconstruct file
"""

import json
import os
import zlib

from sender.protocol import PacketType
from sender.parser import PacketParser


class ReceiverCore:

    def __init__(self):

        self.metadata = None

        self.chunks = {}

        self.total_chunks = 0

        self.transfer_finished = False

    # -------------------------------------------------

    def receive(self, packet_bytes: bytes):

        packet = PacketParser.parse(packet_bytes)

        # CRC Verification
        crc = zlib.crc32(packet.payload)

        if crc != packet.crc32:
            raise ValueError("CRC Check Failed")

        # START
        if packet.packet_type == PacketType.START:

            self.metadata = json.loads(
                packet.payload.decode("utf-8")
            )

            self.total_chunks = self.metadata["total_chunks"]

            print("=" * 50)
            print("TRANSFER STARTED")
            print("=" * 50)

            for key, value in self.metadata.items():
                print(f"{key:15}: {value}")

            print("=" * 50)

            return

        # DATA
        if packet.packet_type == PacketType.DATA:

            self.chunks[packet.chunk_number] = packet.payload

            print(
                f"Received Chunk "
                f"{packet.chunk_number + 1}/"
                f"{self.total_chunks}"
            )

            return

        # END
        if packet.packet_type == PacketType.END:

            print("\nEND Packet Received")

            self.transfer_finished = True

    # -------------------------------------------------

    def completed(self):

        return (
            self.transfer_finished
            and
            len(self.chunks) == self.total_chunks
        )

    # -------------------------------------------------

    def missing_chunks(self):

        missing = []

        for i in range(self.total_chunks):

            if i not in self.chunks:

                missing.append(i)

        return missing

    # -------------------------------------------------

    def save(self, directory="received"):

        if not self.completed():

            raise RuntimeError(
                "Transfer incomplete"
            )

        os.makedirs(
            directory,
            exist_ok=True
        )

        path = os.path.join(
            directory,
            self.metadata["filename"]
        )

        with open(path, "wb") as f:

            for i in range(self.total_chunks):

                f.write(self.chunks[i])

        print("\nSaved:", path)

        return path